File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/analytics/internal/Provider.java
package app.zxtune.analytics.internal;

import android.content.ContentProvider;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.LocaleList;
import android.util.DisplayMetrics;

import androidx.annotation.Nullable;

import java.io.IOException;
import java.util.concurrent.LinkedBlockingQueue;

import app.zxtune.Log;
import app.zxtune.MainApplication;
import app.zxtune.auth.Auth;
import app.zxtune.device.PowerManagement;
import app.zxtune.fs.api.Api;

public final class Provider extends ContentProvider {

  private static final String TAG = Provider.class.getName();

  static final String METHOD_PUSH = "push";

  static final Uri URI = new Uri.Builder().scheme(ContentResolver.SCHEME_CONTENT).authority("app.zxtune.analytics.internal").build();

  private final LinkedBlockingQueue<String> queue = new LinkedBlockingQueue<>(1024);

  @Override
  public boolean onCreate() {
    final Context ctx = getContext();
    if (ctx == null) {
      return false;
    }
    final Context appCtx = ctx.getApplicationContext();
    MainApplication.initialize(appCtx);
    final Auth.UserInfo auth = Auth.getUserInfo(appCtx);
    Api.initialize(auth);
    final UrlsSink delegate = new Dispatcher();
    final Thread thread = new Thread("IASender") {
      @Override
      public void run() {
        try {
          doSending(queue, delegate);
        } catch (Exception e) {
          Log.w(TAG, e, "Stopped IASender thread");
        }
      }
    };
    thread.setDaemon(true);
    thread.start();
    sendSystemInfoEvent();
    sendSystemConfigurationEvent(ctx);
    if (auth.isInitial()) {
      sendInitialInstallationEvent();
    }
    return true;
  }

  private void sendSystemInfoEvent() {
    final UrlsBuilder builder = new UrlsBuilder("system/info");
    builder.addParam("product", Build.PRODUCT);
    builder.addParam("device", Build.DEVICE);
    if (Build.VERSION.SDK_INT >= 21) {
      builder.addParam("cpu_abi", Build.SUPPORTED_ABIS[0]);
      for (int idx = 1; idx < Build.SUPPORTED_ABIS.length; ++idx) {
        builder.addParam("cpu_abi" + (idx + 1), Build.SUPPORTED_ABIS[idx]);
      }
    } else {
      builder.addParam("cpu_abi", Build.CPU_ABI);
      builder.addParam("cpu_abi2", Build.CPU_ABI2);
    }
    builder.addParam("manufacturer", Build.MANUFACTURER);
    builder.addParam("brand", Build.BRAND);
    builder.addParam("model", Build.MODEL);

    builder.addParam("android", Build.VERSION.RELEASE);
    builder.addParam("sdk", Build.VERSION.SDK_INT);

    doPush(builder.getResult());
  }

  private void sendSystemConfigurationEvent(Context ctx) {
    final Resources res = ctx.getResources();
    final UrlsBuilder builder = new UrlsBuilder("system/config");
    final Configuration cfg = res.getConfiguration();
    final DisplayMetrics metrics = res.getDisplayMetrics();
    builder.addParam("layout", cfg.screenLayout);
    if (Build.VERSION.SDK_INT >= 24) {
      final LocaleList locales = cfg.getLocales();
      builder.addParam("locale", locales.get(0).toString());
      for (int idx = 1, lim = locales.size(); idx < lim; ++idx) {
        builder.addParam("locale" + (idx + 1), locales.get(idx).toString());
      }
    } else {
      builder.addParam("locale", cfg.locale.toString());
    }
    builder.addParam("density", metrics.densityDpi);
    builder.addParam("width", cfg.screenWidthDp);
    builder.addParam("height", cfg.screenHeightDp);
    builder.addParam("sw", cfg.smallestScreenWidthDp);
    final Boolean doze = PowerManagement.dozeEnabled(ctx);
    if (doze != null) {
      builder.addParam("doze", doze ? 1 : 0);
    }

    doPush(builder.getResult());
  }

  private void sendInitialInstallationEvent() {
    final UrlsBuilder builder = new UrlsBuilder("app/firstrun");
    // TODO: add installation source
    doPush(builder.getResult());
  }

  @Nullable
  @Override
  public Cursor query(Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
    return null;
  }

  @Nullable
  @Override
  public String getType(Uri uri) {
    return null;
  }

  @Nullable
  @Override
  public Uri insert(Uri uri, @Nullable ContentValues values) {
    return null;
  }

  @Override
  public int delete(Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
    return 0;
  }

  @Override
  public int update(Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
    return 0;
  }

  @Nullable
  @Override
  public Bundle call(String method, @Nullable String arg, @Nullable Bundle extras) {
    if (METHOD_PUSH.equals(method) && arg != null) {
      doPush(arg);
    }
    return null;
  }

  private void doPush(String url) {
    queue.offer(url);
  }

  private static void doSending(LinkedBlockingQueue<String> queue, UrlsSink output) throws InterruptedException, IOException {
    for (; ; ) {
      final String url = queue.take();
      output.push(url);
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/analytics/internal/Provider.kt
package app.zxtune.analytics.internal

import android.content.ContentProvider
import android.content.ContentResolver
import android.content.ContentValues
import android.content.Context
import android.media.AudioManager
import android.media.AudioTrack
import android.net.Uri
import android.os.Build
import android.os.Bundle
import app.zxtune.Logger
import app.zxtune.MainApplication
import app.zxtune.auth.Auth
import app.zxtune.device.PowerManagement.Companion.dozeEnabled
import app.zxtune.fs.api.Api
import app.zxtune.net.NetworkManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class Provider : ContentProvider() {
    @OptIn(ExperimentalCoroutinesApi::class)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO.limitedParallelism(1))
    private lateinit var sink: UrlsSink

    override fun onCreate(): Boolean {
        val ctx = context ?: return false
        val appCtx = ctx.applicationContext
        MainApplication.initialize(appCtx)
        val auth = Auth.getUserInfo(ctx)
        Api.initialize(auth)
        sink = Dispatcher().apply {
            scope.launch {
                NetworkManager.from(appCtx).networkAvailable.collect(this@apply::onNetworkChange)
            }
        }

        sendSystemInfoEvent()
        sendSystemConfigurationEvent(ctx)
        if (auth.isInitial) {
            sendInitialInstallationEvent()
        }
        return true
    }

    override fun shutdown() {
        scope.cancel()
    }

    private fun sendSystemInfoEvent() = UrlsBuilder("system/info").apply {
        addParam("product", Build.PRODUCT)
        addParam("device", Build.DEVICE)
        supportedAbis().forEachIndexed { idx, abi ->
            addIndexedParam("cpu_abi", idx, abi)
        }
        addParam("manufacturer", Build.MANUFACTURER)
        addParam("brand", Build.BRAND)
        addParam("model", Build.MODEL)

        addParam("android", Build.VERSION.RELEASE)
        addParam("sdk", Build.VERSION.SDK_INT.toLong())

        addParam(
            "samplerate", AudioTrack.getNativeOutputSampleRate(AudioManager.STREAM_MUSIC).toLong()
        )
    }.run {
        doPush(result)
    }

    private fun sendSystemConfigurationEvent(ctx: Context) = UrlsBuilder("system/config").apply {
        val res = ctx.resources
        val cfg = res.configuration
        val metrics = res.displayMetrics
        addParam("layout", cfg.screenLayout.toLong())
        if (Build.VERSION.SDK_INT >= 24) {
            cfg.locales.run {
                for (idx in 0..<size()) {
                    addIndexedParam("locale", idx, get(idx).toString())
                }
            }
        } else {
            @Suppress("DEPRECATION") addParam("locale", cfg.locale.toString())
        }
        addParam("density", metrics.densityDpi.toLong())
        addParam("width", cfg.screenWidthDp.toLong())
        addParam("height", cfg.screenHeightDp.toLong())
        addParam("sw", cfg.smallestScreenWidthDp.toLong())
        dozeEnabled(ctx)?.let { doze ->
            addParam("doze", (if (doze) 1 else 0).toLong())
        }
    }.run {
        doPush(result)
    }

    private fun sendInitialInstallationEvent() = UrlsBuilder("app/firstrun").run {
        // TODO: add installation source
        doPush(result)
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ) = null

    override fun getType(uri: Uri) = null

    override fun insert(uri: Uri, values: ContentValues?) = null

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?) = 0

    override fun update(
        uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?
    ) = 0

    override fun call(method: String, arg: String?, extras: Bundle?): Bundle? {
        if (METHOD_PUSH == method && arg != null) {
            doPush(arg)
        }
        return null
    }

    private fun doPush(url: String) = scope.launch {
        runCatching {
            sink.push(url)
        }.onFailure {
            LOG.w(it) { "Failed to push $url" }
        }
    }

    companion object {
        private val LOG = Logger(Provider::class.java.name)

        const val METHOD_PUSH = "push"

        val URI: Uri = Uri.Builder().scheme(ContentResolver.SCHEME_CONTENT)
            .authority("app.zxtune.analytics.internal").build()
    }
}

private fun UrlsBuilder.addIndexedParam(name: String, idx: Int, value: String) =
    addParam(if (idx == 0) name else name + (idx + 1), value)

@Suppress("DEPRECATION")
private fun supportedAbis() = if (Build.VERSION.SDK_INT >= 21) Build.SUPPORTED_ABIS else arrayOf(
    Build.CPU_ABI, Build.CPU_ABI2
)
